### PR TITLE
Notification

### PR DESCRIPTION
**Pull request by synixebrett**
**When merged this pull request will:**
- Create a simple notify function that creates a small notification for mod and mission makers to use rather than using hint or write themselves

### CODE REVIEW

**Comment by jokoho48** [comment]
👎 don't like the use of spawn in there. should use waitAndExecute and waitUntilAndExecute maybe instead?

**Comment by synixebrett** [comment]
I didn't use waitAndExecute since it has multiple blocks requiring sleep. This would require nesting would it not?

I thought the use of 3 times would be better overall then nesting 3 times

**Comment by jokoho48** [comment]
I would build it more like a message queue so that you don't need to wait multiple times for every notification.

so that you have a function that just enqueues the notification, and one that pulls one and draws it and after x amount of time hides it and recalls itself to check if the queue is empty and when then just stops. with that, you only have X "tasks" running at the same time and not a "task" for every notification.

**Comment by synixebrett** [comment]
Wouldn't that require a task running during the entire mission to check the queue? Where as this method has no background tasks running while there a no notifications.

Considering that these notifications would rarely be shown isn't this way better? Even though if 3 notifications are being shown, there are 3 tasks, there should rarely be 3 notifications.

**Comment by jokoho48** [comment]
>Wouldn't that require a task running during the entire mission to check the queue? Where as this method has no background tasks running while there a no notifications.

no, because the function is called when you add a new notification. the draw function/tasks are only running when there is currently a notification on screen or in Queue. but than only the max amount of notifications that can get shown at the same time. and not more.

**Comment by synixebrett** [comment]
> the function is called when you add a new notification

ah I see, I hadn't considered that way

File: addons/ui/fnc_notify.sqf [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+uiNamespace setVariable [QGVAR(NotificationIndex), _index + 1];
+
+waitUntil {_index - GVAR(NotificationCurrentIndex) < GVAR(NotificationMax)};
+
+private _position = _index % GVAR(NotificationMax);
+private _safezoneX = safezoneW + safezoneX;
+
+// Create the 3 displays in the offscreen position
+
+private _ctrlBackground = _display ctrlCreate ["RscPicture", 20];
```
Comment (commy2) [inline addons/ui/fnc_notify.sqf]: Magic variable should be a macro.

File: addons/ui/fnc_notify.sqf [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+_ctrlBackground ctrlSetPosition [
+    _safezoneX,
+    (0.24 + (_position * 0.08)) * safezoneH + (safezoneY * -0.5) ,
+    0.1 * safezoneW,
+    0.06 * safezoneH
+];
+_ctrlBackground ctrlSetText "#(argb,8,8,3)color(0,0,0,0.75)";
```
Comment (commy2) [inline addons/ui/fnc_notify.sqf]: I feel like this should use ctrlSetBackgroundColor instead. This as picture control seems awkward to me at least.

File: addons/ui/fnc_notify.sqf [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+_ctrlBackground ctrlCommit 0;
+
+private _ctrlTitle = _display ctrlCreate ["RscStructuredText", 10];
```
Comment (commy2) [inline addons/ui/fnc_notify.sqf]: Magic variable should be a macro.

File: addons/ui/fnc_notify.sqf [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    _safezoneX,
+    (0.24 + (_position * 0.08)) * safezoneH + (safezoneY * -0.5),
+    0.1 * safezoneW,
+    0.0255 * safezoneH
+];
+_ctrlTitle ctrlSetStructuredText parseText _title;
+_ctrlTitle ctrlSetTextColor _color;
+_ctrlTitle ctrlCommit 0;
+
+private _ctrlNotification = _display ctrlCreate ["RscStructuredText", 30];
```
Comment (commy2) [inline addons/ui/fnc_notify.sqf]: Magic variable should be a macro.

File: addons/ui/fnc_notify.sqf [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+_ctrlNotification ctrlSetPosition
+[
```
Comment (commy2) [inline addons/ui/fnc_notify.sqf]: Inconsistent formatting. I prefer L57 & L67, and that is the style used in CBA in most places.

File: addons/ui/fnc_notify.sqf [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+// move the displays onscreen
+
+private _ctrls = [_ctrlBackground, _ctrlTitle, _ctrlNotification];
+{
+    private _position = ctrlPosition _x;
+    private _posX = 0.9 * safezoneW + safezoneX;
+    _position set [0, _posX];
+    _x ctrlSetPosition _position;
+    _x ctrlCommit (uiNamespace getVariable [QGVAR(NotificationSpeed), 0.5]);
```
Comment (commy2) [inline addons/ui/fnc_notify.sqf]: Why are these ui namespace? Doesn't settings put the variables in mission namespace? It seems weird to put this into ui namespace.

**Comment by commy2** [comment]
Not happy about this being scheduled. Basically breaks if you save and load a savegame or have poor FPS in MP.

**Comment by commy2** [comment]
This should be a controls group control template in config, so the script doesn't have to recreate the whole thing with slow sqf every time.

**Comment by dedmen** [comment]
arma3_x64 2018-11-30 23-23-26-06
Screenshot from @commy2 
It's isn't of much use if it's in slack and not here ;)
The above obviously won't get merged like that

**Comment by synixebrett** [comment]
That's what it looks like?

Wildly different from my PC.

I am going to be rewriting this PR after my exams anyway.

**Comment by commy2** [comment]
I have a counter proposal here: #1036, which is kind of a port from the ACE one.

**Pull request closed without merging**